src/backend/distributed/transaction/transaction_recovery.c

```diff
@@ -182,191 +184,221 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
 
-	int subtractIndex = 0;
-	int subtractCount = list_length(subtractList);
-	char **subtractArray = (char **) PointerArrayFromList(subtractList);
+		hash_search(recheckTransactionSet, transactionName, HASH_FIND,
+					&foundPreparedTransactionAfterCommit);
 
-	foreach(nameCell, nameList)
-	{
-		char *baseName = (char *) lfirst(nameCell);
+		if (foundPreparedTransactionBeforeCommit && foundPreparedTransactionAfterCommit)
+		{
+			/*
+			 * The transaction was committed, but the prepared transaction still exists
+			 * on the worker. Try committing it.
+			 *
+			 * We double check that the recovery record exists both before and after
+			 * checking ActiveDistributedTransactionNumbers(), since we may have
+			 * observed a prepared transaction that was committed immediately after.
+			 */
+			bool shouldCommit = true;
+			bool commitSucceeded = RecoverPreparedTransactionOnWorker(connection,
+																	  transactionName,
+																	  shouldCommit);
+			if (!commitSucceeded)
+			{
+				/*
+				 * Failed to commit on the current worker. Stop without throwing
+				 * an error to allow recover_prepared_transactions to continue with
+				 * other workers.
+				 */
+				recoveryFailed = true;
+				break;
+			}
+
+			recoveredTransactionCount++;
 
-		bool nameFound = FindMatchingName(subtractArray, subtractCount,
-										  baseName, &subtractIndex);
+			/*
+			 * We successfully committed the prepared transaction, safe to delete
+			 * the recovery record.
+			 */
+		}
+		else if (foundPreparedTransactionAfterCommit)
+		{
+			/*
+			 * We found a committed pg_dist_transaction record that initially did
+			 * not have a prepared transaction, but did when we checked again.
+			 *
+			 * If a transaction started and committed just after we observed the
+			 * set of prepared transactions, and just before we called
+			 * ActiveDistributedTransactionNumbers, then we would see  a recovery
+			 * record without a prepared transaction in pendingTransactionSet,
+			 * but there may be prepared transactions that failed to commit.
+			 * We should not delete the records for those prepared transactions,
+			 * since we would otherwise roll back them on the next call to
+			 * recover_prepared_transactions.
+			 *
+			 * In addition, if the transaction started after the call to
+			 * ActiveDistributedTransactionNumbers and finished just before our
+			 * pg_dist_transaction snapshot, then it may still be in the process
+			 * of comitting the prepared transactions in the post-commit callback
+			 * and we should not touch the prepared transactions.
+			 *
+			 * To handle these cases, we just leave the records and prepared
+			 * transactions for the next call to recover_prepared_transactions
+			 * and skip them here.
+			 */
 
-		if (!nameFound)
+			continue;
+		}
+		else
 		{
 			/*
-			 * baseName is not in subtractArray and thus included
-			 * in the difference.
+			 * We found a recovery record without any prepared transaction. It
```

Comment:
Is that ever possible `(!foundPreparedTransactionAfterCommit && !foundPreparedTransactionBeforeCommit)`? I think not, but just wanted to ask.

Reply:
Yes, that's actually the common case. Any 2PC that successfully completed before the call to `recover_prepared_transactions()` will have a `pg_dist_transaction` record, but no prepared transaction, since it was already committed. In that case we just delete the record.